tool/tbot/identity/identity.go

```diff
@@ -0,0 +1,388 @@
+		return nil, trace.BadParameter("Cert: missing parameter")
+	}
+
+	cert, err := apisshutils.ParseCertificate(certBytes)
+	if err != nil {
+		return nil, trace.BadParameter("failed to parse server certificate: %v", err)
+	}
+
+	signer, err := ssh.ParsePrivateKey(keyBytes)
+	if err != nil {
+		return nil, trace.BadParameter("failed to parse private key: %v", err)
+	}
+	// this signer authenticates using certificate signed by the cert authority
+	// not only by the public key
+	certSigner, err := ssh.NewCertSigner(cert, signer)
+	if err != nil {
+		return nil, trace.BadParameter("unsupported private key: %v", err)
+	}
+
+	// check principals on certificate
+	if len(cert.ValidPrincipals) < 1 {
+		return nil, trace.BadParameter("valid principals: at least one valid principal is required")
+	}
+	for _, validPrincipal := range cert.ValidPrincipals {
+		if validPrincipal == "" {
+			return nil, trace.BadParameter("valid principal can not be empty: %q", cert.ValidPrincipals)
+		}
+	}
+
+	// check permissions on certificate
+	// TODO: do we care to even verify this at all?
+	// if len(cert.Permissions.Extensions) == 0 {
+	// 	return nil, trace.BadParameter("extensions: misssing needed extensions for host roles")
+	// }
+
+	// roleString := cert.Permissions.Extensions[utils.CertExtensionRole]
+	// if roleString == "" {
+	// 	return nil, trace.BadParameter("misssing cert extension %v", utils.CertExtensionRole)
+	// }
+	// roles, err := types.ParseTeleportRoles(roleString)
+	// if err != nil {
+	// 	return nil, trace.Wrap(err)
+	// }
+	// foundRoles := len(roles)
+	// if foundRoles != 1 {
+	// 	return nil, trace.Errorf("expected one role per certificate. found %d: '%s'",
+	// 		foundRoles, roles.String())
+	// }
+
+	// TODO: host certs use CertExtensionAuthority, client certs use CertExtensionTeleportRouteToCluster
+	// (or at least, teleport-route-to-cluster _appears_ to be a sane cluster name, but I have no idea
+	// how it is actually set.)
+	//clusterName := cert.Permissions.Extensions[utils.CertExtensionAuthority]
+	clusterName := cert.Permissions.Extensions[teleport.CertExtensionTeleportRouteToCluster]
+	if clusterName == "" {
+		return nil, trace.BadParameter("missing cert extension %v", utils.CertExtensionAuthority)
+	}
+
+	return &Identity{
+		ClusterName:       clusterName,
+		KeyBytes:          keyBytes,
+		SSHPublicKeyBytes: publicKeyBytes,
+		CertBytes:         certBytes,
+		KeySigner:         certSigner,
+		Cert:              cert,
+	}, nil
+}
+
+func SaveIdentity(id *Identity, d destination.Destination) error {
+	for _, data := range []struct {
+		name     string
+		data     []byte
+		modeHint destination.ModeHint
+	}{
+		{TLSCertKey, id.TLSCertBytes, destination.ModeHintSecret},
+		{SSHCertKey, id.CertBytes, destination.ModeHintSecret},
+		{TLSCACertsKey, bytes.Join(id.TLSCACertsBytes, []byte("$")), destination.ModeHintSecret},
+		{SSHCACertsKey, bytes.Join(id.SSHCACertBytes, []byte("\n")), destination.ModeHintSecret},
```

Comment:
any rationale for the different separators? I blindly changed both to `"$"` [here](https://github.com/gravitational/teleport/pull/10397) because SSH CA certs seem to have a terminal newline that was messing up the splitter

Reply:
Not especially, nope. If that's simpler I say just switch it.